## DATA TRAINING DAN TESTING

***Melakukan pemodelan data dengan regresi  untuk memporoleh data dengan dengan proporsi 70:30 (70% untuk data testing)*** 

### *Aulia Nur Fitriani*
### *21083010051*
### *Machine Learning I (C)*

In [2]:
pwd

'C:\\Users\\USER\\Documents\\machine learning 1'

In [3]:
#import library pandas dan numpy untuk analisis data
import pandas as pd
import numpy as np

In [4]:
Nama_Variabel=['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'MEDV']
dataset = pd.read_csv('housing.csv', header = None, delim_whitespace=True, names=Nama_Variabel)

# Menampilkan Data
print(dataset)

        CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...    ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273.0   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273.0   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273.0   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273.0   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273.0   

     PTRATIO       B  LSTAT  MEDV  
0       15.3  396.90   4.98  24.0  
1       17.8  396.90   9.14

In [5]:
# untuk mengetahui 5 data teratas
dataset.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

In [6]:
# untuk mengetahui banyaknya jumlah kolom dan baris
np.shape(dataset)

(506, 14)

In [7]:
# untuk mengetahui tipe data
dataset.dtypes

CRIM       float64
ZN         float64
INDUS      float64
CHAS         int64
NOX        float64
RM         float64
AGE        float64
DIS        float64
RAD          int64
TAX        float64
PTRATIO    float64
B          float64
LSTAT      float64
MEDV       float64
dtype: object

In [8]:
# untuk mengubah data kategori pada kolom "CHAS"
dataset["CHAS"] = dataset["CHAS"].astype('category')

dataset.dtypes

CRIM        float64
ZN          float64
INDUS       float64
CHAS       category
NOX         float64
RM          float64
AGE         float64
DIS         float64
RAD           int64
TAX         float64
PTRATIO     float64
B           float64
LSTAT       float64
MEDV        float64
dtype: object

In [9]:
# mengetahui perhitungan statistik seperti informasi jumlah data (count), rata-rata (mean), standar deviasi (std), nilai minimum (min), kuartil (25%, 50%, dan 75%), dan nilai maksimum (max) dari setiap kolom pada DataFrame atau Series.
dataset.describe()

CRIM          ZN       INDUS         NOX          RM         AGE  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.554695    6.284634   68.574901   
std      8.601545   23.322453    6.860353    0.115878    0.702617   28.148861   
min      0.006320    0.000000    0.460000    0.385000    3.561000    2.900000   
25%      0.082045    0.000000    5.190000    0.449000    5.885500   45.025000   
50%      0.256510    0.000000    9.690000    0.538000    6.208500   77.500000   
75%      3.677082   12.500000   18.100000    0.624000    6.623500   94.075000   
max     88.976200  100.000000   27.740000    0.871000    8.780000  100.000000   

              DIS         RAD         TAX     PTRATIO           B       LSTAT  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.795043    9.549407  408.237154   18.455534  356.674032   12.653063   
std      2.105710    8.707259  168.537116    2.164946   91.294864    7.141062   
min      1.129600    1.000000  187.000000   12.600000    0.320000    1.730000   
25%      2.100175    4.000000  279.000000   17.400000  375.377500    6.950000   
50%      3.207450    5.000000  330.000000   19.050000  391.440000   11.360000   
75%      5.188425   24.000000  666.000000   20.200000  396.225000   16.955000   
max     12.126500   24.000000  711.000000   22.000000  396.900000   37.970000   

             MEDV  
count  506.000000  
mean    22.532806  
std      9.197104  
min      5.000000  
25%     17.025000  
50%     21.200000  
75%     25.000000  
max     50.000000

In [10]:
# menghitung jumlah kemunculan setiap nilai unik pada kolom 'CHAS' 
dataset['CHAS'].value_counts()

0    471
1     35
Name: CHAS, dtype: int64

In [11]:
# memeriksa apa ada data yang missing value
dataset.isnull().any()

# berdasarkan hasil dibawah ini, menunjukkan bahwa tidak ada data yang missing value pada setiap kolom

CRIM       False
ZN         False
INDUS      False
CHAS       False
NOX        False
RM         False
AGE        False
DIS        False
RAD        False
TAX        False
PTRATIO    False
B          False
LSTAT      False
MEDV       False
dtype: bool

In [12]:
# Melakukan transformasi min-max scaling pada data
from sklearn import preprocessing
transformasi_min_max = preprocessing.MinMaxScaler()

***Melakukan pemodelan data dengan regresi*** 

In [13]:
# Variabel target diisi dengan nama kolom yang akan dijadikan target variabel dalam analisis, yaitu 'MEDV'.
target = 'MEDV'
x = dataset.drop('MEDV', axis =1) # metode .drop() pada DataFrame dataset dan menghapus kolom 'MEDV'
y = dataset[target]
varx = ['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT']
x = pd.DataFrame(data = transformasi_min_max.fit_transform(x), columns = varx)

In [14]:
# train_test_split() digunakan untuk membagi data menjadi dua bagian yaitu data latih (train set) dan data uji (test set) dalam proporsi tertentu.
# dilakukan pembagian data dengan proporsi 70:30 (70% untuk data testing) menggunakan fungsi train_test_split(). 

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.70, random_state = 1)

In [15]:
# dilakukan pemodelan data dengan metode regresi linear
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

linear_model = LinearRegression()
linear_model.fit(x_train, y_train)

LinearRegression()

In [16]:
# dilakukan transformasi data polinomial 
from sklearn.preprocessing import PolynomialFeatures
polynomial_features = PolynomialFeatures(degree = 2)
x_poly = polynomial_features.fit_transform(x_train)
poly_model = LinearRegression()
poly_model.fit(x_poly, y_train)
x_poly_test = polynomial_features.fit_transform(x_test)

In [17]:
#Pemodelan dengan Regresi Ridge dengan nilai alpha 1

from sklearn.linear_model import Ridge

model_ridge = Ridge(alpha = 1)
model_ridge.fit(x_train, y_train)

Ridge(alpha=1)

In [18]:
#Pemodelan dengan Regresi Lasso dengan nilai alpha 1

from sklearn.linear_model import Lasso

model_lasso = Lasso(alpha = 1)
model_lasso.fit(x_train, y_train)

Lasso(alpha=1)

### *Melakukan Evaluasi Data Testing dan Training dengan  Regresi*

***Melakukan Evaluasi Data Testing dan Testing dengan Model Regresi Linear***

In [19]:
# untuk mengetahui nilai RMSE dan Rsquare(R2) pada data training dalam metode regresi linear
pred_y_train = linear_model.predict(x_train)
rmse = (np.sqrt(mean_squared_error(y_train, pred_y_train)))
r2 = r2_score(y_train, pred_y_train)

print("====================================")
print("Performa Model Data Training yaitu :")
print("--------------------------------------")
print("NILAI RMSE -> {}".format(rmse))
print("NILAI R2 -> {}".format(r2))

Performa Model Data Training yaitu :
--------------------------------------
NILAI RMSE -> 4.980768789994864
NILAI R2 -> 0.7254441891569579


In [20]:
# untuk mengetahui nilai RMSE dan Rsquare(R2) pada data testing dalam metode regresi linear
pred_y_test = linear_model.predict(x_test)
rmse_2 = (np.sqrt(mean_squared_error(y_test, pred_y_test)))
r2_2 = r2_score(y_test, pred_y_test)

print("===================================")
print("Performa Model Data Testing yaitu :")
print("--------------------------------------")
print("NILAI RMSE -> {}".format(rmse_2))
print("NILAI R2 -> {}".format(r2_2))

Performa Model Data Testing yaitu :
--------------------------------------
NILAI RMSE -> 4.877665892510709
NILAI R2 -> 0.7094540602303248


***Melakukan Evaluasi Data Testing dan Training dengan Model Regresi Polinomial***

In [21]:
# untuk mengetahui nilai RMSE dan Rsquare(R2) pada data training dalam metode regresi polinomial
pred_y_train_poly = poly_model.predict(x_poly)
rmse = (np.sqrt(mean_squared_error(y_train, pred_y_train_poly)))
r2 = r2_score(y_train, pred_y_train_poly)

print("===================================")
print("Performa Model Polinomial Derajat Dua Data Training yaitu :")
print("--------------------------------------")
print("NILAI RMSE -> {}".format(rmse_2))
print("NILAI R2 -> {}".format(r2_2))

Performa Model Polinomial Derajat Dua Data Training yaitu :
--------------------------------------
NILAI RMSE -> 4.877665892510709
NILAI R2 -> 0.7094540602303248


In [22]:
# untuk mengetahui nilai RMSE dan Rsquare(R2) pada data testing dalam metode regresi polinomial
pred_y_test_poly = poly_model.predict(x_poly_test)
rmse_2 = (np.sqrt(mean_squared_error(y_test, pred_y_test_poly)))
r2_2 = r2_score(y_test, pred_y_test_poly)

print("===================================")
print("Performa Model Polinomial Derajat Dua Data Testing yaitu :")
print("--------------------------------------")
print("NILAI RMSE -> {}".format(rmse_2))
print("NILAI R2 -> {}".format(r2_2))

Performa Model Polinomial Derajat Dua Data Testing yaitu :
--------------------------------------
NILAI RMSE -> 12.417733962960714
NILAI R2 -> -0.8831088112505843


***Melakukan Evaluasi Data Testing dan Training dengan Model Regresi Ridge***

In [23]:
# untuk mengetahui nilai RMSE dan Rsquare(R2) pada data training dalam metode regresi ridge
pred_y_train_ridge = model_ridge.predict(x_train)
rmse = (np.sqrt(mean_squared_error(y_train, pred_y_train_ridge)))
r2 = r2_score(y_train, pred_y_train_ridge)

print("===================================")
print("Performa Model Ridge Data Training yaitu :")
print("--------------------------------------")
print("NILAI RMSE -> {}".format(rmse_2))
print("NILAI R2 -> {}".format(r2_2))

Performa Model Ridge Data Training yaitu :
--------------------------------------
NILAI RMSE -> 12.417733962960714
NILAI R2 -> -0.8831088112505843


In [24]:
# untuk mengetahui nilai RMSE dan Rsquare(R2) pada data testing dalam metode regresi ridge
pred_y_test_ridge = model_ridge.predict(x_test)
rmse_2 = (np.sqrt(mean_squared_error(y_test, pred_y_test_ridge)))
r2_2 = r2_score(y_test, pred_y_test_ridge)

print("===================================")
print("Performa Model Ridge Data Testing yaitu :")
print("--------------------------------------")
print("NILAI RMSE -> {}".format(rmse_2))
print("NILAI R2 -> {}".format(r2_2))

Performa Model Ridge Data Testing yaitu :
--------------------------------------
NILAI RMSE -> 4.7808135250163595
NILAI R2 -> 0.7208778371246569


***Melakukan Evaluasi Data Testing dan Training dengan Model Regresi Lasso***

In [25]:
# untuk mengetahui nilai RMSE dan Rsquare(R2) pada data training dalam metode regresi lasso
pred_y_train_lasso = model_lasso.predict(x_train)
rmse = (np.sqrt(mean_squared_error(y_train, pred_y_train_lasso)))
r2 = r2_score(y_train, pred_y_train_lasso)

print("===================================")
print("Performa Model Lasso Data Training yaitu :")
print("--------------------------------------")
print("NILAI RMSE -> {}".format(rmse_2))
print("NILAI R2 -> {}".format(r2_2))

Performa Model Lasso Data Training yaitu :
--------------------------------------
NILAI RMSE -> 4.7808135250163595
NILAI R2 -> 0.7208778371246569


In [26]:
# untuk mengetahui nilai RMSE dan Rsquare(R2) pada data testing dalam metode regresi lasso
pred_y_test_lasso = model_lasso.predict(x_test)
rmse_2 = (np.sqrt(mean_squared_error(y_test, pred_y_test_lasso)))
r2_2 = r2_score(y_test, pred_y_test_lasso)

print("===================================")
print("Performa Model Lasso Data Testing yaitu :")
print("--------------------------------------")
print("NILAI RMSE -> {}".format(rmse_2))
print("NILAI R2 -> {}".format(r2_2))

Performa Model Lasso Data Testing yaitu :
--------------------------------------
NILAI RMSE -> 7.707774930476879
NILAI R2 -> 0.27448104703573784


 ***Dalam evaluasi kualitas model regresi, R-square dan RMSE adalah metrik umum yang digunakan. R-square mengukur proporsi variabilitas dalam data yang dapat dijelaskan oleh model regresi, sementara RMSE mengukur kesalahan prediksi model dalam satuan yang sama dengan data yang diamati. Model regresi yang baik memberikan nilai R-square yang tinggi dan nilai RMSE yang rendah. Dalam kasus tertentu, Model Polynomial dipilih sebagai model terbaik karena memiliki nilai R-Square dan RMSE yang baik, namun pemilihan model terbaik tidak hanya bergantung pada R-square dan RMSE saja tetapi juga harus mempertimbangkan faktor lain seperti akurasi, kecepatan, dan kompleksitas model.***
 
***Berdasarkan ke empat metode untuk mendapatkan nilai RMSE dan Rsquare(R2) didapatkan hasil dengan metode terbaik yaitu pada metode "Model Regresi Ridge" dengan nilai RMSE paling rendah dan nilai Rsquare(R2) paling tinggi yaitu "4.7808135250163595"
dan "0.7208778371246569" dibandingkan dengan hasil metode yang lainnya***